content/blog/fast-data-processing-pipeline-for-predicting-flight-delays-using-apache-apis-kafka-spark-streaming-and-machine-learning-part-3.md

```diff
@@ -0,0 +1,295 @@
+    "dest": "SFO",
+    "crsdephour": 11,
+    "crsdeptime": 1120.0,
+    "crsarrtime": 1308.0,
+    "crselapsedtime": 168.0,
+    "pred_dtree": 1.0,
+    "pred_randforest": 1.0,
+    "pred_svm": 1.0,
+    "actual_delay": 1.0
+
+
+}
+```
+
+MapR Event Store allows processing of the same messages by different consumers. This makes it easy to add different consumers for the same message. With this type of architecture and flexible schema, you can easily add and deploy new microservices with new machine learning models.
+
+**Spark Streaming writing to MapR Database**
+
+The MapR Database OJAI Connector for Apache Spark enables you to use [MapR Database as a sink for Apache Spark Data Streams](https://docs.datafabric.hpe.com/62/Spark/SavingDStreamMapRDB.html).
+
+![Spark Streaming writing to MapR Database](/img/spark-streaming-writing-to-mapr-db.png)
+
+You can read about the MapR Event Streams Spark Streaming code in [part 2 of this series](https://developer.hpe.com/blog/fast-data-processing-pipeline-for-predicting-flight-delays-using-apache-apis-kafka-spark-streaming-and-machine-learning-part-2/); here, we will focus on Spark streaming writing to MapR Database. The messages from the MapR Database topic are in JSON format and contain the following for each flight: the flight id, day of the week, carrier, origin, destination, scheduled departure hour, scheduled departure time, scheduled arrival time, scheduled travel time, delay prediction, and actual delay label (Note in the real world architecture the actual delay label would come later in a different topic, but to keep the tutorial code simple it is combined here). An example is shown below:
+
+```json
+{
+    "_id": "UA_2017-03-28_DEN_SFO_721",
+    "dofW": 2,
+    "carrier": "UA",
+    "origin": "DEN",
+    "dest": "SFO",
+    "crsdephour": 11,
+    "crsdeptime": 1120.0,
+    "crsarrtime": 1308.0,
+    "crselapsedtime": 168.0,
+    "label": 0.0,
+    "pred_dtree": 1.0
+}
+```
+
+Below we use a Scala case class and [Structype](https://spark.apache.org/docs/latest/sql-programming-guide.html#programmatically-specifying-the-schema) to define the schema, corresponding to the input data.
+
+![Scala case class 1](/img/scala-case-class.png)
+
+![Scala case class 2](/img/scala-case-class-2.png)
+
+We use the KafkaUtils createDirectStream method with Kafka configuration parameters to create an input stream from a MapR Event Store topic. This creates a DStream that represents the stream of incoming data, where each message is a key value pair. We use the DStream map transformation to create a DStream with the message values.
+
+![KafkaUtils createDirectStream method](/img/kafkautils-createdirectstream-method.png)
+
+![DStream](/img/dstream.png)
+
+In the code below each RDD in the valuesDStream is transformed into a Spark Dataset. Then the MapR Database Spark Connector DStream saveToMapRDB method performs a parallel partitioned bulk insert of JSON FlightwPred objects into MapR Database.
```

Comment:
Add commas after “below” and “Then”

![image](https://user-images.githubusercontent.com/77496261/137183424-990d260f-2ca4-4b0a-9f99-752ee485c5c4.png)

Reply:
done